### PR TITLE
Add the memtrack AIDL HAL implement

### PR DESCRIPTION
**Pull request by yhe39**
Tracked-On: OAM-102243
Signed-off-by: HeYue <yue.he@intel.com>

### CODE REVIEW

**Comment by yhe39** [comment]
Hi @JeevakaPrabu @feijiang1 @YuanjunHuang, could you help to review this patch?
This patch implement the AIDL Memtrack HAL.

**Review by YuanjunHuang — APPROVED** [review]
LGTM, thanks

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
++
++#include <hardware/hardware.h>
++#include <hardware/memtrack.h>
++
+ namespace aidl {
+ namespace android {
+ namespace hardware {
+ namespace memtrack {
++namespace impl {
++namespace example {
```
Comment (feijiang1) [inline aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:52]: why need the namespace example?

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
++        delete[] legacy_records;
++    }
++
+     return ndk::ScopedAStatus::ok();
+ }
+ 
+ ndk::ScopedAStatus Memtrack::getGpuDeviceInfo(std::vector<DeviceInfo>* _aidl_return) {
+     _aidl_return->clear();
+-    DeviceInfo dev_info = {.id = 0, .name = "virtio_gpu"};
++    DeviceInfo dev_info = {.id = 0, .name = "i915"};
```
Comment (feijiang1) [inline aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:103]: does it mean only mentrack only exist for CiV GVT-d and SRIOV?

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+         return ndk::ScopedAStatus(AStatus_fromExceptionCode(EX_UNSUPPORTED_OPERATION));
+     }
+     _aidl_return->clear();
++
++    size_t temp = 0;
++    size_t *size = &temp;
++    int ret = 0;
++
++    if (mModule->getMemory == nullptr)
++    {
```
Comment (feijiang1) [inline aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:73]: coding style is not consistent. Other places use
if () {
}

**Review by feijiang1 — COMMENTED** [review]
please see my comments

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch (hunk 1 above)
Comment (yhe39) [reply aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:52]: i checked the power and light aidl implementation, they all use the impl and example name space.
The memtrack service name is android.hardware.memtrack-service.example
So i add it there.

Comment (feijiang1) [reply aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:52]: AOSP provide the example, but our implementation should not be example anymore.

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch (hunk 2 above)
Comment (yhe39) [reply aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:103]: In fact this line: DeviceInfo dev_info = {.id = 0, .name = "i915"}; is needed for pass the VTS test.
Before someone add the virtio_gpu name to pass the VTS test.
I change it to i915 as we implement the AIDL hal and kernel have the memtrack function patch.

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch (hunk 1 above)
Comment (yhe39) [reply aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:52]: The light aidl hal is implemented by Yuanzhe before, he also use the example, you could check here:
https://github.com/projectceladon/vendor-intel-utils/pull/1070/files
I could remove the example namespace, i just follow the other aild implementation examples.

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch (hunk 3 above)
Comment (yhe39) [reply aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:73]: OK, i will modify later :)

File: aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch (hunk 2 above)
Comment (yhe39) [reply aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch:103]: i have checked on virtio, there is no kernel related memtrack function enabled.
Although the hal service is working, but there is no memtrack EGL info:
caas:/ # dumpsys meminfo |grep EGL                                             
          0K: EGL mtrack
          0K: EGL mtrack
caas:/ # service list |grep mem
10	android.hardware.memtrack.IMemtrack/default: [android.hardware.memtrack.IMemtrack]
119	meminfo: []
120	memtrack.proxy: [android.hardware.memtrack.IMemtrack]
caas:/ #

**Review by JeevakaPrabu — APPROVED** [review]

**Review by feijiang1 — APPROVED** [review]

**Review by feijiang1 — APPROVED** [review]

**Review by JeevakaPrabu — APPROVED** [review]

**Pull request merged**